File: leetcode/leet88.py
# py2
# merge-sorted-array

class Solution(object):
  def merge(self, nums1, m, nums2, n):
    """
    :type nums1: List[int]
    :type nums2: List[int]
    :type m: int
    :type n: int
    :rtype: None Do not return anything, modify nums1 in-place instead.
    """
    list1 = nums1[:m]
    list2 = nums2[:n]
    i, j = 0, 0
    nums1[:] = []
    # print list1, list2
    while i < len(list1) and j < len(list2):
      # list1[i] > list2[j]
      if list1[i] > list2[j]:
        nums1.append(list2[j])
        j += 1
      else:
        nums1.append(list1[i])
        i += 1
      # print nums1
    if i < len(list1):
      nums1[i+j:] = list1[i:]
    if j < len(list2):
      nums1[i+j:] = list2[j:]
    print nums1

if __name__ == "__main__":
  testData = [
    [[1,2,3,0,0,0], 3, [2,5,6], 3]
  ]
  test = Solution()
  for i in testData:
    print i
    test.merge(i[0], i[1], i[2], i[3])

File: leetcode/level1/leet88.js
/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
var merge = function(nums1, m, nums2, n) {
    let index1 = m - 1
    let index2 = n - 1
    let index3 = nums1.length - 1
    while (index3 > -1) {
        if (index2 < 0) {
            return nums1
        }

        if (index1 < 0) {
            while (index2 > -1) {
                nums1[index3] = nums2[index2]
                index2--
                index3--
            }
            return nums1
        }

        if (nums1[index1] > nums2[index2]) {
            nums1[index3] = nums1[index1]
            index1--
        } else {
            nums1[index3] = nums2[index2]
            index2--
        }
        index3--
    }
    return nums1
};

var testData = [
    // { nums1: [1,2,3,0,0,0], m: 3, nums2: [2,5,6], n: 3},
    { nums1: [2,0], m: 1, nums2: [1], n: 1},
    // { nums1: [0], m: 0, nums2: [1], n: 1},
]

for (let v of testData) {
    console.log(merge(v.nums1, v.m, v.nums2, v.n))
}